### ISSUE TITLE
CrossEntropyLoss mishandles weights

### ISSUE THREAD

**Issue by bhancock8** [pytorch/pytorch#8062]
## Issue description
When a `weight` tensor is passed to CrossEntropyLoss, it exhibits unusual behavior, such as occasionally ignoring the weights, and returning different results when it averages losses over examples versus averaging them manually.

## Code example
1) Incorrect size_average result
```
input = torch.tensor([
    [-100.,  100.],
    [-100.,  100.]
])
target = torch.tensor([0,1], dtype=torch.long)
weights = torch.tensor([20,10], dtype=torch.float)

ce_noreduce = nn.CrossEntropyLoss(weight=weights, reduce=False)
ce_reduce = nn.CrossEntropyLoss(weight=weights, reduce=True, size_average=True)

loss1 = ce_noreduce(input, target)
loss2 = ce_reduce(input, target)

print(loss1)
print(loss1.mean())
print(loss2)
```
output:
```
tensor([ 4000.,    -0.])
tensor(2000.)
tensor(133.3333)
```

2) Weights ignored
```
input = torch.tensor([
    [-1,  1, -1],
    [-1,  1, -1],
    [-1,  1, -1],
], dtype=torch.float) * 100
target = torch.tensor([0,0,1], dtype=torch.long)
weights = torch.tensor([1,10,1], dtype=torch.float)

ce_noweights = nn.CrossEntropyLoss(weight=None, reduce=False)
ce_weights = nn.CrossEntropyLoss(weight=weights, reduce=False)

loss1 = ce_noweights(input, target)
loss2 = ce_weights(input, target)

print(loss1.mean())
print(loss2.mean())
```
output:
```
tensor([ 0,  0,  1], dtype=torch.uint8)
tensor(133.3333)
tensor(133.3333)
```

## System Info
```
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.4
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] msgpack-numpy (0.4.1)
[pip3] numpy (1.13.3)
[pip3] numpydoc (0.7.0)
[pip3] torch (0.3.0.post4)
[pip3] torchfile (0.1.0)
[pip3] torchnet (0.0.1)
[pip3] torchtext (0.2.1)
[pip3] torchvision (0.2.0)
[conda] pytorch                   0.4.0           py36_cuda0.0_cudnn0.0_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
```

**Comment by bhancock8** [comment]
I just confirmed that I get the same results using pytorch 0.3.0.

**Comment by li-roy** [comment]
I think the output of both of your scripts is actually expected behavior. Let me know if I'm misunderstanding or if you disagree.

In the first example, we're take an average weighted on the weights array, so (4000 + 0)/(20 + 10) = 133.333 which is expected.

In the second example, I think you're not noticing a change due to weights because the third loss element ends up being zero, and it's the only one that has label 1 (thus the only one that would be affected by the 10 in your weights array). If you change the first element of your weights array, you'll be able to see a change reflected in your loss:

```
input = torch.tensor([
    [-1,  1, -1],
    [-1,  1, -1],
    [-1,  1, -1],
], dtype=torch.float) * 100
target = torch.tensor([0,0,1], dtype=torch.long)
weights = torch.tensor([10,10,1], dtype=torch.float)

ce_noweights = nn.CrossEntropyLoss(weight=None, reduce=False)
ce_weights = nn.CrossEntropyLoss(weight=weights, reduce=False)

loss1 = ce_noweights(input, target)
loss2 = ce_weights(input, target)

print(loss1.mean())
print(loss2.mean())
```
output:
```
tensor(133.3333)
tensor(1333.3334)
```

**Comment by bhancock8** [comment]
Thanks for taking a look, @li-roy. Looking at the first example:

It seems that there is a disconnect between the documentation and the implementation. The docs (https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss) make no mention of dividing by the sum of the weights; they simply show the loss for each class being multiplied by its weight, which is the behavior I expected.

It also seems surprising that applying the size-average operation manually (by passing reduce=False and then doing a .mean() operation) does not return the same result as passing reduce=True and size_average=True. This could be rectified by either dividing each individual example's loss by the weights as you suggested (even when reduce=False), or being consistent with the documentation as mentioned above and simply using the weights as raw multipliers. Let me know if you disagree. Thanks!

**Comment by li-roy** [comment]
Sorry for the late response. We'll discuss this and get back to you!

**Comment by li-roy** [comment]
We recently reworked size_average and reduce to a single reduction argument. I think the right way to approach this is to add additional options in reduction to specify a weighted vs unweighted average.
